lightrec.c

```diff
@@ -199,44 +199,48 @@ static void lightrec_invalidate_map(struct lightrec_state *state,
 }
 
 static const struct lightrec_mem_map *
-lightrec_get_map(struct lightrec_state *state, u32 kaddr)
+lightrec_get_map(struct lightrec_state *state,
+		 void **host, u32 kaddr)
 {
+	static const struct lightrec_mem_map *map;
 	unsigned int i;
 
 	for (i = 0; i < state->nb_maps; i++) {
-		const struct lightrec_mem_map *map = &state->maps[i];
+		const struct lightrec_mem_map *mapi = &state->maps[i];
 
-		if (kaddr >= map->pc && kaddr < map->pc + map->length)
-			return map;
+		if (kaddr >= mapi->pc && kaddr < mapi->pc + mapi->length) {
+			map = mapi;
+			break;
+		}
 	}
 
-	return NULL;
+	if(i == state->nb_maps)
+		return NULL;
+
+	while (map->mirror_of)
+		map = map->mirror_of;
+
+	*host = map->address + kaddr - map->pc;
```

Comment:
You could check `if (host)` first, then callers could pass NULL if they are not interested in the host address.